dsgrid/config/dimension_associations.py

```diff
@@ -172,3 +172,21 @@ def iter_associations(self):
         """Yields the stored dimension types with their association tables as a tuple."""
         for dimensions, table in self._associations.items():
             yield dimensions, table
+
+    def get_full_join_by_data_source(self, data_source):
```

Comment:
@elainethale @mooneyme @lixiliu I think that producing a single table of dimension associations is something like this. I'm unsure about `how="cross"` vs `how="outer"`. Results appear the same in limited testing.

I think that we could implement all checks by keeping this separate from the full combinations of project records. We could still iterate over pairs of combinations in project_registry_manager and check if they exist here. If there was somehow implicit relationships that still need to be accounted for, that wouldn't work.

Let me know what you think. If this seems right to you then I can implement the logic today. Thanks.

Reply:
@daniel-thom here is what I have been playing around with. Given our current test, this works as we would expect. However I think there are a few case study tests we need to test out to make sure it behaves as expected under every potential scenario.

```
    def get_full_join_by_data_source(self, project_config, data_source):
        tables = []
        for dims, val in self._associations.items():
            if DimensionType.DATA_SOURCE in dims:
                continue
            table = self.get_associations_by_data_source(data_source, *dims)
            tables.append(table)

        table = tables[0]
        if len(tables) > 1:
            for i, other in enumerate(tables[1:]):
                print(i)
                on_columns = list(set(other.columns).intersection(table.columns)) # TODO: is there potentially an issue here with the ordering of the tables to lead to a "non-overlap" on_columns?
                if on_columns:
                    table_columns = table.columns
                    table_distinct_cols_count =  table.select(table.columns).distinct().count() 
                    table = table.join(other, on=on_columns, how="outer")
                    assert table.select(other.columns).distinct().count() == other.select(other.columns).distinct().count() 
                    assert table.select(table_columns).distinct().count() == table_distinct_cols_count 
                else:
                    # TODO: need to test this with the right combinations of supplied associations
                    table = table.crossJoin(other)
                    
        # TODO: in our current test using standard scenarios dimensions, we only have 5 dimensions in the columns, but 8 tables are joined together. This has to do with the full circle chain of associations provided. We may be able to save time and skip these as long as we do some sanity checks?? Unclear ATM.
        
        all_dimensions = set(d.value for d in DimensionType)
        missing_dimensions = all_dimensions.difference(table.columns)
        for dim in missing_dimensions:
            if dim not in ("data_source", "time"): # TODO: how do we want to add in time dimension if we don't have records for them?
                table_count = table.count()
                # TODO: @dtom is there a better/preferred way to get the records of a dimension?
                other = project_config.get_base_dimension(DimensionType(dim)).get_records_dataframe().select("id as geography")
                table = other.crossJoin(table.withColumnRenamed("id", dim))
                assert table.count() == other.count() * table.count()
        
        table = table.withColumn("data_source", lit(data_source))
        return table
```

Reply:
Also just a note. I found that it is an "outer join" when combining association tables provided that have an overlapping dimension column, however it is a "cross join" when we are joining association tables without an overlapping column or dimensions that were not provided with an association table (e.g., geography).